Mock api server for weshlist

weirdjh opened the pull request:
## Api 
###  Channel API
- GET /channel/{channelName}/join
- GET /channel/{channelId}/quit ... Not implemented
- GET /channel/{channelId}/music/search ... Not implemented
- GET /channel/{channelId}/playlist

### Music API
- POST /music/add

### User API
- POST /user/{userId}/music

## MongoDB
Run Embedded Mongo DB on local environment and initialize some mock data using it.
```
- user0 ~ user4
- music0 ~ music7
- channel0  ~ channel1
```
## ROP (Railway Oriented Programming)
Trying to programming with
```
/**
 * A result of a computation that can succeed or fail.
 */
sealed class Result<out T, out E> {
	data class Success<out T>(val value: T) : Result<T, Nothing>()
	data class Failure<out E>(val reason: E) : Result<Nothing, E>()
}
```
which makes failsafe program.

## TODO
- Clean up logging
- Write test code
- Write docs with Swagger

Commit:
Init project with room-mocking api server

weirdjh committed:
Error propagate handling

weirdjh committed:
Add validate logic

weirdjh committed:
.DS_Store banished!

weirdjh committed:
Add logging

weirdjh committed:
Add embedded mongoDB

weirdjh committed:
Get playlist from channel

weirdjh committed:
Attach java logger to embedded mongodb

tlsqlfhdnsdl commented on api/src/main/kotlin/io/weshlist/minutemaid/result/Result.kt:
```diff
@@ -0,0 +1,76 @@
+fun <T> Result<T, T>.get() = when (this) {
+	is Result.Success<T> -> value
+	is Result.Failure<T> -> reason
+}
```
Result<T, E> 가 아닌 Result<T, T> 인 이유는 어떤 건가요?

weirdjh replied:
고치는 게 맞아보입니다 ^^

weirdjh replied:
<T, E>로 고치는게 맞아보입니다ㅎㅎ

tlsqlfhdnsdl commented on api/src/main/kotlin/io/weshlist/minutemaid/utils/Response.kt:
```diff
@@ -0,0 +1,30 @@
+			error = reason
+		)
+	}
+}
```
api response 에 대해 같이 얘기를 나눠보고싶은게 있어요 ㅎㅎ

tlsqlfhdnsdl reviewed (commented):
훌륭합니다! 덕분에 코틀린에 대한 이해도가 쑥쑥 올라가는 기분이에요

tlsqlfhdnsdl commented on api/src/main/kotlin/io/weshlist/minutemaid/controller/ChannelController.kt:
```diff
@@ -0,0 +1,106 @@
+	@GetMapping("/{channelName}/join")
+	fun join(
+		@PathVariable channelName: String,
+		@RequestParam userId: UserID
```
GetMapping 으로 하신 이유가 궁금합니다! join 은 결국 resource 를 생성할 수 있는 행위이기에 post 가 적절하다고 생각했어요

weirdjh replied:
아 지난번에 얘기를 했었는데요, 제가 고치는걸 깜빡했습니다. 수정하겠습니다.

weirdjh committed:
Apply code review

1. Change channel join http method.
2. Fix result get method.
3. Remove result code in RestApiResponse.
